src/java.base/share/classes/java/lang/Integer.java

```diff
@@ -961,7 +961,16 @@ private static final class IntegerCache {
             if (archivedCache == null || size > archivedCache.length) {
                 Integer[] c = new Integer[size];
                 int j = low;
-                for(int i = 0; i < c.length; i++) {
+                // Use all cached values from the archive to avoid breaking
```

Comment:
Alternatively, you could use a separate array to store **only** the runtime created cached instances, without copying the references from the archived cache. Both the archived cache and runtime added cache are used. That has slightly less memory and efficiency overhead, but almost unnoticeable.

Reply:
I don't think I understand. The whole point is to make sure the runtime cache instances are the same as archive cache instances, right? Using both archived and runtime caches leads to identity inconsistencies: if we `==` compare the archived `Integer` box and the runtime `Integer` box, they should equal each other. So we have to make sure we use the parts we have already committed to use (from the archive), and add whatever cache elements are missing at runtime. I don't see how it works otherwise.

Reply:
The runtime cache would only store the additional boxed Integers not in the archived cache, without affecting the archived boxed Integers. Something like the following would work (I didn't test all corner cases):

```
            static final int archivedSize;
            ...
            archivedSize = archivedCache == null ? 0 : archivedCache.length;
            if (archivedCache == null || size > archivedCache.length) {
                int runtimeCacheSize = size - archivedSize;
                Integer[] c = new Integer[runtimeCacheSize];
                int j = low + archivedSize;
                for(int i = 0; i < runtimeCacheSize; i++) {
                    c[i] = new Integer(j++);
                }

                cache = c;
                if (archivedCache == null) {
                    archivedCache = c;
                }
            } else {
                cache = null;
            }
```

`Integer.valueOf()` would retrieve the cached boxed Integer either from the archived cache or the runtime created cache, if the value is within the runtime [low, high].

```
    public static Integer valueOf(int i) {
        if (i >= IntegerCache.low && i <= IntegerCache.high) {
            int index = i + (-IntegerCache.low);
            if (IntegerCache.archivedSize != 0 && IntegerCache.archivedSize >= index) {
                return IntegerCache.archivedCache[index];
            } else {
                return IntegerCache.cache[index - IntegerCache.archivedSize];
            }      
        }
        return new Integer(i); 
    }
```

Reply:
Ah, I see. I don't think we want to do this, though. I vaguely remember `Integer.valueOf` code shape being important for things like C2 optimizations around EA and autoboxing elimination. The beauty of staying within a single array is avoiding a whole class of these issues.

Reply:
I think it's mostly ok if we don't take the archivedCache+runtimeCache approach. The overhead is not too large in common cases, as likely the archivedCache would have the default size if runtime changes to use a different `high` value.